### PR TITLE
rbd-mirror: add support for snapshot-based mirroring resyncs

### PR DESCRIPTION
**Pull request by dillaman**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/librbd/mirror/snapshot/ImageMeta.cc [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  } else if (r < 0) {
+    lderr(m_image_ctx->cct) << "failed to load snapshot-based mirroring "
+                            << "image-meta: " << cpp_strerror(r) << dendl;
+    on_finish->complete(r);
+    return;
+  }
+
+  bool json_valid = false;
+  json_spirit::mValue json_root;
+  if(json_spirit::read(data, json_root)) {
```
Comment (trociny) [inline src/librbd/mirror/snapshot/ImageMeta.cc]: nit: missing whitespace

**Comment by dillaman** [comment]
> It seems I don't see where and when the image meta resync flag is reset to false, after the resync is complete?

The local image is deleted upon resync so it isn't required

> And I think we would need to filter out the resync meta keys in image metadata API requests?

I can do that, but it technically isn't required since I filter it out later in an attempt to handle it all in the update metadata state.

> And when mirroring is disabled we need to clean the image meta?

Yes, I should probably delete the key.

**Comment by dillaman** [comment]
update pushed

**Comment by trociny** [comment]
@dillaman It needs rebase after #33440 was merged.

**Comment by trociny** [comment]
> The local image is deleted upon resync so it isn't required

Indeed :-)

> I can do that, but it technically isn't required since I filter it out later in an attempt to handle it all in the update metadata state.

I mean, is it ok that users would see the key when running `rbd image-meta list` and can update/remove it? We could just make invisible all keys with ".rbd" or "." prefix and forbid users to update that namespace.

I seemed to have a patch for this in one of my previous "mirror snapshot" versions, but it was dropped as not necessary at that time.

**Comment by dillaman** [comment]
> I mean, is it ok that users would see the key when running `rbd image-meta list` and can update/remove it? We could just make invisible all keys with ".rbd" or "." prefix and forbid users to update that namespace.

I don't really mind the user seeing these keys if interested. I should, however, probably make sure we don't copy the key to clones and copied images, though. Let me append one more commit.

**Comment by trociny** [comment]
> I don't really mind the user seeing these keys if interested.

Theoretically, although extremely unlikely, a user may create ".rbd_mirror.{uuid}" key and break mirroring. So, a valid (allowed) user action leads to an invalid situation.

But this is a really artificial case, and I am fine with any solution, jut brought your attention.

**Comment by dillaman** [comment]
> But this is a really artificial case, and I am fine with any solution, jut brought your attention.

Yeah, understood.

**Comment by dillaman** [comment]
updated

**Comment by dillaman** [comment]
jenkins test make check

**Comment by dillaman** [comment]
unittest_rbd_mirror failure w/ cut-off logs: https://jenkins.ceph.com/job/ceph-pull-requests/45598/console

Jenkins has been updated to not reboot on failures now so retrying build to see if we can see log output.

File: src/librbd/mirror/snapshot/SetImageStateRequest.cc [hunk 2]
```diff
@@ -77,50 +78,30 @@ void SetImageStateRequest<I>::handle_get_snap_limit(int r) {
 
 template <typename I>
 void SetImageStateRequest<I>::get_metadata() {
-  CephContext *cct = m_image_ctx->cct;
-  ldout(cct, 20) << "start_key=" << m_last_metadata_key << dendl;
-
-  librados::ObjectReadOperation op;
-  cls_client::metadata_list_start(&op, m_last_metadata_key, MAX_METADATA_ITEMS);
+   CephContext *cct = m_image_ctx->cct;
```
Comment (trociny) [inline src/librbd/mirror/snapshot/SetImageStateRequest.cc]: nit: seems like an extra space

Comment (dillaman) [reply src/librbd/mirror/snapshot/SetImageStateRequest.cc]: fixed, thx.

File: src/tools/rbd_mirror/image_replayer/snapshot/PrepareReplayRequest.cc [hunk 3]
```diff
@@ -22,12 +23,41 @@ namespace mirror {
...
+  load_local_image_meta();
+}
+
+template <typename I>
+void PrepareReplayRequest<I>::load_local_image_meta() {
+  dout(15) << dendl;
+
+  ceph_assert(m_state_builder->local_image_meta == nullptr);
+  m_state_builder->local_image_meta = new librbd::mirror::snapshot::ImageMeta(
+    m_state_builder->local_image_ctx, m_local_mirror_uuid);
```
Comment (trociny) [inline src/tools/rbd_mirror/image_replayer/snapshot/PrepareReplayRequest.cc]: @dillaman shaman failed to build it [1]:
```
CMakeFiles/rbd_mirror_internal.dir/image_replayer/snapshot/PrepareReplayRequest.cc.o -c /build/ceph-15.1.0-1216-g86045ba/src/tools/rbd_mirror/image_replayer/snapshot/PrepareReplayRequest.cc
/build/ceph-15.1.0-1216-g86045ba/src/tools/rbd_mirror/image_replayer/snapshot/PrepareReplayRequest.cc: In member function 'void rbd::mirror::image_replayer::snapshot::PrepareReplayRequest<ImageCtxT>::load_local_image_meta()':
/build/ceph-15.1.0-1216-g86045ba/src/tools/rbd_mirror/image_replayer/snapshot/PrepareReplayRequest.cc:42:58: error: class template argument deduction failed:
     m_state_builder->local_image_ctx, m_local_mirror_uuid);
                                                          ^
/build/ceph-15.1.0-1216-g86045ba/src/tools/rbd_mirror/image_replayer/snapshot/PrepareReplayRequest.cc:42:58: error: no matching function for call to 'ImageMeta()'
In file included from /build/ceph-15.1.0-1216-g86045ba/src/tools/rbd_mirror/image_replayer/snapshot/PrepareReplayRequest.cc:10:0:
/build/ceph-15.1.0-1216-g86045ba/src/librbd/mirror/snapshot/ImageMeta.h:27:3: note: candidate: template<class ImageCtxT> ImageMeta(ImageCtxT*, const string&)-> librbd::mirror::snapshot::ImageMeta<ImageCtxT>
   ImageMeta(ImageCtxT* image_ctx, const std::string& mirror_uuid);
   ^~~~~~~~~
/build/ceph-15.1.0-1216-g86045ba/src/librbd/mirror/snapshot/ImageMeta.h:27:3: note:   template argument deduction/substitution failed:
/build/ceph-15.1.0-1216-g86045ba/src/tools/rbd_mirror/image_replayer/snapshot/PrepareReplayRequest.cc:42:58: note:   candidate expects 2 arguments, 0 provided
     m_state_builder->local_image_ctx, m_local_mirror_uuid);
                                                          ^
src/tools/rbd_mirror/CMakeFiles/rbd_mirror_internal.dir/build.make:1241: recipe for target 'src/tools/rbd_mirror/CMakeFiles/rbd_mirror_internal.dir/image_replayer/snapshot/PrepareReplayRequest.cc.o' failed
```
[1] https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=bionic,DIST=bionic,MACHINE_SIZE=huge/37446//consoleFull

**Comment by dillaman** [comment]
Fixed [1]

[1] https://shaman.ceph.com/builds/ceph/wip-43934/

File: src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc [hunk 4]
```diff
@@ -898,74 +930,134 @@ void Replayer<I>::handle_unlink_peer(int r) {
...
-  refresh_local_image();
+  load_local_image_meta();
 }
 
 template <typename I>
-void Replayer<I>::register_update_watcher() {
+void Replayer<I>::register_local_update_watcher() {
   dout(10) << dendl;
 
   m_update_watch_ctx = new C_UpdateWatchCtx(this);
```
Comment (trociny) [inline src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc]: @dillaman The memcheck reports error:
```
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758== 352 bytes in 22 blocks are definitely lost in loss record 7 of 11
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    at 0x4C31556: operator new(unsigned long) (vg_replace_malloc.c:344)
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x56CC80: rbd::mirror::image_replayer::snapshot::Replayer<librbd::ImageCtx>::register_local_update_watcher() (Replayer.cc:940)
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x56D224: rbd::mirror::image_replayer::snapshot::Replayer<librbd::ImageCtx>::init(Context*) (Replayer.cc:157)
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x492F34: rbd::mirror::ImageReplayer<librbd::ImageCtx>::start_replay() (ImageReplayer.cc:417)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x4931D4: rbd::mirror::ImageReplayer<librbd::ImageCtx>::handle_bootstrap(int) (ImageReplayer.cc:402)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x49B6FF: complete (Context.h:77)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x49B6FF: rbd::mirror::CancelableRequest::finish(int) (CancelableRequest.h:30)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x512216: rbd::mirror::image_replayer::BootstrapRequest<librbd::ImageCtx>::close_remote_image() (BootstrapRequest.cc:422)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x513A54: rbd::mirror::image_replayer::BootstrapRequest<librbd::ImageCtx>::handle_prepare_replay(int) (BootstrapRequest.cc:320)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x565843: complete (Context.h:77)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x565843: finish (BaseRequest.h:22)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x565843: rbd::mirror::image_replayer::snapshot::PrepareReplayRequest<librbd::ImageCtx>::handle_load_local_image_meta(int) (PrepareReplayRequest.cc:62)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x7E25F7: complete (Context.h:77)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x7E25F7: librbd::mirror::snapshot::ImageMeta<librbd::ImageCtx>::handle_load(Context*, int) (ImageMeta.cc:71)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x7DF99B: complete (Context.h:77)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x7DF99B: void librbd::util::detail::rados_callback<LambdaContext<librbd::mirror::snapshot::ImageMeta<librbd::ImageCtx>::load(Context*)::{lambda(int)#2}> >(void*, LambdaContext<librbd::mirror::snapshot::ImageMeta<librbd::ImageCtx>::load(Context*)::{lambda(int)#2}>) (Utils.h:27)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x4EE02A0: librados::C_AioComplete::finish(int) (AioCompletionImpl.h:140)
```
We use this variable for both local and remote watch contexts and delete only once.

[1] http://qa-proxy.ceph.com/teuthology/trociny-2020-02-26_17:11:11-rbd-wip-43934-distro-basic-smithi/4805442/teuthology.log

Comment (dillaman) [reply src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc]: Oops -- definitely only need a single update watch context.

Comment (dillaman) [reply src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc]: Deleted the second `new` operation in `register_remote_update_watcher` below.

**Pull request merged**